### DISCUSSION TITLE
Scaled grid item with fixed width

### DISCUSSION THREAD

**Discussion by Abdullahbutt3434** [react-grid-layout/react-grid-layout#2046] [Q&A]
I am making a dashboard builder which have lots of widget. each widget can have tables chart etc. I want to make responsive in a way that if I open the dashboard on mobile screen the grid item should retain the width and some out everything like the video below

https://github.com/react-grid-layout/react-grid-layout/assets/48261058/5e49ef5c-319b-4f1e-ac19-9a12dd4a61b7

In my scenario this is happening

https://github.com/react-grid-layout/react-grid-layout/assets/48261058/b856d44c-0993-4a89-9986-57998ce24f29

Anyone any idea how I can achieve this zoomOut and stop changing in actual width of widget.
here is my main component of Grid Layout. I tried to implement scale but this is not working

function ResponsiveGridLayout(props) {
  const {
    isDraggable = true,
    isResizable = true,
    rowHeight = 60,
    className = 'react-grid-layout layout',
    breakpoints = { lg: 1600 },
    cols = { lg: 12 },
    draggableHandle = '.drag-handle',
    onLayoutChangeHandle = () => {},
    handleOnDragStop = () => {},
    handleOnResizeStop = () => {},
    onBreakpointChangeHandle = () => {},
    sidebarWidth = 250 // Add this prop for sidebar width
  } = props;

  const [scale, setScale] = useState(1);

  // Setting up the WidthProvider
  const GridLayout = useMemo(() => WidthProvider(Responsive), []);

  const updateScale = () => {
    const mainContainer = document.querySelector('.main');
    if (mainContainer) {
      const containerWidth = mainContainer.offsetWidth;
      const availableWidth = window.innerWidth - 280; // Subtract sidebar width from available width
      const newScale = availableWidth / containerWidth;
      setScale(newScale);
    }
  };

  useEffect(() => {
    updateScale();
    window.addEventListener('resize', updateScale);
    return () => window.removeEventListener('resize', updateScale);
  }, []);

  return (
      <div className={"main"} style={{
        transform: `scale(${scale})`,
        transformOrigin: '0 0',
        width: '100%',
        height: '100%',
        overflow: 'hidden'
      }}>
        <GridLayout
            isDraggable={isDraggable}
            isResizable={isResizable}
            rowHeight={rowHeight}
            className={className}
            breakpoints={breakpoints}
            cols={cols}
            draggableHandle={draggableHandle}
            onLayoutChange={onLayoutChangeHandle}
            onBreakpointChange={onBreakpointChangeHandle}
            onDragStop={handleOnDragStop}
            onResizeStop={handleOnResizeStop}
        >
          {props.children}
        </GridLayout>
      </div>
  );
}

export default ResponsiveGridLayout;

**Comment by amanyadav-work** [comment]
You can try this out, Add a ref to parent section where you're rendering the layout and then pass row height to Grid Layout.

`   const [rowHeight, setRowHeight] = useState(null);
    const sectionRef = useRef(null);

    useEffect(() => {
        const element = sectionRef.current;
        if (!element) return;

        // Measure immediately on mount
        setRowHeight(element.getBoundingClientRect().width * 0.3);

        const observer = new ResizeObserver(([entry]) => {
            const calculatedWidth = Math.min(entry.contentRect.width * 0.3, 400); // Adjust accordingly
            setRowHeight(calculatedWidth);
        });

        observer.observe(element);

        return () => observer.disconnect();
    }, []);
`

`   <ReactGridLayout
      className="layout react-draggable "
      cols={3}
       rowHeight={rowHeight} // Pass here
       layout={layout}
       onLayoutChange={handleLayoutChange}
       draggableCancel=".no-drag"
>`


Worked for me.

**Comment by davideagosti-dev** [comment]
Hi @Abdullahbutt3434 — I came across this while looking specifically at responsive dashboard builders with resizable widgets.

You mentioned having many different widgets, including tables and charts, and trying to preserve their behavior across different screen sizes. That’s very close to the problem space I’m working on with DnDGem.

DnDGem is an early-alpha open-source layout library built around the idea that “a component geometrically fits” and “its content is still useful at that size” are two different things.

If you’re still working on this dashboard builder, I’d be very interested in understanding how your widgets behave across those size changes. If there’s a fit, I’d be happy to help you try DnDGem against the real use case and work through any gaps we find.

https://github.com/davideagosti-dev/dndgem
